### PR TITLE
GDScript Documentation

### PR DESCRIPTION
**Pull request by martindzejky**
I wrote the descriptions for the built-in gdscript functions. The only one remaining is the `str2var`. I couldn't figure out from Godot nor the source code, what it does. Any ideas?

### CODE REVIEW

**Comment by akien-mga** [comment]
Apart from my comments on the first commit, the changes are good. However, could you merge all the additions into one commit? Having so many commits just for little additions will make the git history harder to read for no real gain (also harder to cherry-pick changes).

**Comment by chuckeles** [comment]
Working on it... Fix the first commit and then squash. Go it!

**Comment by chuckeles** [comment]
Is that better?

**Comment by akien-mga** [comment]
That's better yes. I'm still puzzled by the random reordering of constants that were previously sorted by enum value, and are no longer. I'll see if I can reproduce that in master HEAD too.

**Comment by akien-mga** [comment]
I ran doctool myself and had only clean changes:
```
diff --git a/doc/base/classes.xml b/doc/base/classes.xml
index e658e2a..74bb26b 100644
--- a/doc/base/classes.xml
+++ b/doc/base/classes.xml
@@ -13897,6 +13897,108 @@ Example: (content-length:12), (Content-Type:application/json; charset=UTF-8)
 		</constant>
 	</constants>
 </class>
+<class name="HTTPRequest" inherits="Node" category="Core">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<methods>
+		<method name="request">
+			<return type="int">
+			</return>
+			<argument index="0" name="url" type="String">
+			</argument>
+			<argument index="1" name="custom_headers" type="StringArray" default="StringArray()">
+			</argument>
+			<argument index="2" name="ssl_validate_domain" type="bool" default="true">
+			</argument>
+			<description>
+			</description>
+		</method>
+		<method name="cancel_request">
+			<description>
+			</description>
+		</method>
+		<method name="get_http_client_status" qualifiers="const">
+			<return type="int">
+			</return>
+			<description>
+			</description>
+		</method>
+		<method name="set_use_threads">
+			<argument index="0" name="enable" type="bool">
+			</argument>
+			<description>
+			</description>
+		</method>
+		<method name="is_using_threads" qualifiers="const">
+			<return type="bool">
+			</return>
+			<description>
+			</description>
+		</method>
+		<method name="set_body_size_limit">
+			<argument index="0" name="bytes" type="int">
+			</argument>
+			<description>
+			</description>
+		</method>
+		<method name="get_body_size_limit" qualifiers="const">
+			<return type="int">
+			</return>
+			<description>
+			</description>
+		</method>
+		<method name="set_max_redirects">
+			<argument index="0" name="amount" type="int">
+			</argument>
+			<description>
+			</description>
+		</method>
+		<method name="get_max_redirects" qualifiers="const">
+			<return type="int">
+			</return>
+			<description>
+			</description>
+		</method>
+	</methods>
+	<signals>
+		<signal name="request_completed">
+			<argument index="0" name="result" type="int">
+			</argument>
+			<argument index="1" name="response_code" type="int">
+			</argument>
+			<argument index="2" name="headers" type="StringArray">
+			</argument>
+			<argument index="3" name="body" type="RawArray">
+			</argument>
+			<description>
+			</description>
+		</signal>
+	</signals>
+	<constants>
+		<constant name="RESULT_SUCCESS" value="0">
+		</constant>
+		<constant name="RESULT_CHUNKED_BODY_SIZE_MISMATCH" value="1">
+		</constant>
+		<constant name="RESULT_CANT_CONNECT" value="2">
+		</constant>
+		<constant name="RESULT_CANT_RESOLVE" value="3">
+		</constant>
+		<constant name="RESULT_CONNECTION_ERROR" value="4">
+		</constant>
+		<constant name="RESULT_SSL_HANDSHAKE_ERROR" value="5">
+		</constant>
+		<constant name="RESULT_NO_RESPONSE" value="6">
+		</constant>
+		<constant name="RESULT_BODY_SIZE_LIMIT_EXCEEDED" value="7">
+		</constant>
+		<constant name="RESULT_REQUEST_FAILED" value="8">
+		</constant>
+		<constant name="RESULT_REDIRECT_LIMIT_REACHED" value="9">
+		</constant>
+	</constants>
+</class>
 <class name="HingeJoint" inherits="Joint" category="Core">
 	<brief_description>
 	</brief_description>
@@ -17285,6 +17387,56 @@ Example: (content-length:12), (Content-Type:application/json; charset=UTF-8)
 	<constants>
 	</constants>
 </class>
+<class name="LinkButton" inherits="BaseButton" category="Core">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<methods>
+		<method name="set_text">
+			<argument index="0" name="text" type="String">
+			</argument>
+			<description>
+			</description>
+		</method>
+		<method name="get_text" qualifiers="const">
+			<return type="String">
+			</return>
+			<description>
+			</description>
+		</method>
+		<method name="set_underline_mode">
+			<argument index="0" name="underline_mode" type="int">
+			</argument>
+			<description>
+			</description>
+		</method>
+		<method name="get_underline_mode" qualifiers="const">
+			<return type="int">
+			</return>
+			<description>
+			</description>
+		</method>
+	</methods>
+	<constants>
+		<constant name="UNDERLINE_MODE_ALWAYS" value="0">
+		</constant>
+		<constant name="UNDERLINE_MODE_ON_HOVER" value="1">
+		</constant>
+	</constants>
+	<theme_items>
+		<theme_item name="underline_spacing" type="int">
+		</theme_item>
+		<theme_item name="font_color" type="Color">
+		</theme_item>
+		<theme_item name="font_color_hover" type="Color">
+		</theme_item>
+		<theme_item name="font_color_pressed" type="Color">
+		</theme_item>
+		<theme_item name="font" type="Font">
+		</theme_item>
+	</theme_items>
+</class>
 <class name="MainLoop" inherits="Object" category="Core">
 	<brief_description>
 	Main loop is the abstract main loop base class.
```

**Comment by akien-mga** [comment]
Actually, if I can bother you some more: it would be nice to have one first commit that just uses doctool to sync the docs (and thus add the new HTTPRequest stuff), and another commit with your own additions. This way I can cherry-pick the second commit for the stable 2.0 branch, while the first commit would not be relevant there.

**Comment by chuckeles** [comment]
Shouldn't I then just remove the doctool changes and only leave my own changes?

**Comment by chuckeles** [comment]
Also, do you know what `str2var` does? I would love to also add that so that it's complete

**Comment by neikeq** [comment]
`var2str` uses _VariantWriter_ and `str2var` uses _VariantParser_. This means you can convert strings returned by `var2str` back to _Variant_ using `str2var`.

**Comment by trollworkout** [comment]
@chuckeles str2var turns a string into a Variant
str2var("4.3") will become var 4.3 (float)

**Comment by akien-mga** [comment]
> Shouldn't I then just remove the doctool changes and only leave my own changes?

You could do that too. If you can investigate why your doctool changes messed up the ordering of some constants though, it might come in handy for the next time :)

**Comment by trollworkout** [comment]
Also thank you. Is about time someone started adding the missing class description to the classes.xml.

**Comment by akien-mga** [comment]
> Is about time someone started adding the missing class description to the classes.xml.

It's not like we've been idle on this, just check the git history for classes.xml ;)

File: doc/base/classes.xml [hunk 1]
```diff
@@ -604,6 +607,7 @@
 			<argument index="3" name="a8" type="int">
 			</argument>
 			<description>
+			Makes a color from red, gree, blue and alpha.
```
Comment (eska014) [inline doc/base/classes.xml]: Also typo "gree"

Comment (eska014) [inline doc/base/classes.xml]: Color8 is specifically for constructing a Color from 8-bit values (0-255), in contrast to the Color constructor (0.0-1.0)

**Comment by chuckeles** [comment]
So that's why I couldn't figure out why this doesn't work:

sv1
sv2

So it needs to be like this (because that's what var2str returns):

sv3

Got it. Will update the docs.

File: doc/base/classes.xml [hunk 2]
```diff
@@ -619,6 +623,7 @@
 			<argument index="0" name="instance_id" type="int">
 			</argument>
 			<description>
+			Get an object by it's ID.
```
Comment (eska014) [inline doc/base/classes.xml]: "its" rather than "it's"

**Comment by trollworkout** [comment]
@chuckeles  var2str needs a Variant object not a string object "chuckeles" is not a Variant
```
var string = "chuckeles"
var string2 = var2str(string) #valid
```

**Comment by neikeq** [comment]
@trollworkout AFAIK, there is no difference between your example and his. Both are String Variants. Am I missing something? :P

**Comment by chuckeles** [comment]
@neikeq No, there's not. It needs to be `"\"chuckeles\"", I understand it now.

File: doc/base/classes.xml (hunk 2 above)
Comment (chuckeles) [inline doc/base/classes.xml]: oh, this always gets me -_-

**Comment by neikeq** [comment]
I mean the following:
``` GDScript
var string = "chuckeles"
print(var2str(string))
```
should be the same as:
``` GDScript
print(var2str("chuckeles"))
```

**Comment by trollworkout** [comment]
From what I know "hello" is a C++ string object while var stringVar = "hello" is a Variant object. So there is some extra magic happening in var2str where any value you give it is first converted to a Variant object then parsed.

File: doc/base/classes.xml [hunk 3]
```diff
@@ -42939,4 +42948,4 @@ This method controls whether the position between two cached points is interpola
 	<constants>
 	</constants>
 </class>
-</doc>
+</doc>
```
Comment (chuckeles) [inline doc/base/classes.xml]: I have no idea how this got here or how to remove this change, none of my editors detect it :disappointed: If it's a problem, let me know

**Comment by chuckeles** [comment]
I finally cleaned the commit to only include my changes! Phew, that took a lot of work.

File: doc/base/classes.xml (hunk 3 above)
Comment (eska014) [inline doc/base/classes.xml]: You're missing an empty newline at the end of the file

Comment (chuckeles) [inline doc/base/classes.xml]: All right fixed

**Comment by akien-mga** [comment]
Thanks :)

**Pull request merged**